text/0000-io-safety.md

```diff
@@ -0,0 +1,429 @@
+/// doesn't have a way to reassign or independently drop it.
+unsafe impl IoSafe for MyType {}
+```
+
+It requires `unsafe`, to require the code to explicitly commit to upholding I/O
+safety. With `IoSafe`, the `do_some_io` example should simply add a
+`+ IoSafe` to provide I/O safety:
+
+```rust
+pub fn do_some_io<FD: AsRawFd + IoSafe>(input: &FD) -> io::Result<()> {
+    some_syscall(input.as_raw_fd())
+}
+```
+
+## Gradual adoption
+
+I/O safety and `IoSafe` wouldn't need to be adopted immediately, adoption
+could be gradual:
+
+ - First, `std` adds `IoSafe` with impls for all the relevant `std` types.
+   This is a backwards-compatible change.
+
+ - After that, crates could implement `IoSafe` for their own types. These
+   changes would be small and semver-compatible, without special coordination.
+
+ - Once the standard library and enough popular crates utilize `IoSafe`,
+   crates could start to add `+ IoSafe` bounds (or adding `unsafe`), at their
+   own pace. These would be semver-incompatible changes, though most users of
+   APIs adding `+ IoSafe` wouldn't need any changes.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## The I/O safety concept
+
+In addition to the Rust language's memory safety, Rust's standard library also
+guarantees I/O safety. An I/O operation is *valid* if the raw handles
+([`RawFd`], [`RawHandle`], and [`RawSocket`]) it operates on are values
+explicitly returned from the OS, and the operation occurs within the lifetime
+the OS associates with them. Rust code has *I/O safety* if it's not possible
+for that code to cause invalid I/O operations.
+
+While some OS's document their file descriptor allocation algorithms, a handle
+value predicted with knowledge of these algorithms isn't considered "explicitly
+returned from the OS".
+
+Functions accepting arbitrary raw I/O handle values ([`RawFd`], [`RawHandle`],
+or [`RawSocket`]) should be `unsafe` if they can lead to any I/O being
+performed on those handles through safe APIs.
+
+Functions accepting types implementing
+[`AsRawFd`]/[`IntoRawFd`]/[`AsRawHandle`]/[`AsRawSocket`]/[`IntoRawHandle`]/[`IntoRawSocket`]
+should add a `+ IoSafe` bound if they do I/O with the returned raw handle.
+
+## The `IoSafe` trait
+
+Types implementing `IoSafe` guarantee that they uphold I/O safety. They must
+not make it possible to write a safe function which can perform invalid I/O
+operations, and:
+
+ - A type implementing `AsRaw* + IoSafe` means its `as_raw_*` function returns
+   a handle which is valid to use for the duration of the `&self` reference.
+   If such types have methods to close or reassign the handle without
+   dropping the whole object, they must document the conditions under which
+   existing raw handle values remain valid to use.
+
+ - A type implementing `IntoRaw* + IoSafe` means its `into_raw_*` function
+   returns a handle which is valid to use at the point of the return from
+   the call.
```

Comment:
Could you give an example of a type, other than RawFd or a constructed equivalent, which would not satisfy this property?

Reply:
So, in practice, almost any non-broken impl of AsRawFd should implement this, and it primarily excludes RawFd itself?

Reply:
Yes.